phac-nml/irida-next
# Attachments Concatenation Service

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds in the service layer to concatenate single and paired end reads. Single end files can only be concatenated with other single end files, and paired end files can only be concatenated with other paired end files. The service accepts params in the following formats

# single-end files
1) attachment_ids is just a 1 dimensional array and is required
2) basename is the base name for the file
3) delete_originals can be set to true to delete the original files. This is also optional

`params: { attachment_ids: [], basename: "FILE BASENAME", delete_originals: true }`

# paired-end files
1) attachment_ids is just a 2 dimensional array and is required. Each array should be a set of paired-end files
2) basename is the base name for the file
3) delete_originals can be set to true to delete the original files. This is also optional

`params: { attachment_ids: [[],[],...], basename: "FILE BASENAME", delete_originals: true }`

Addresses #238  

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

N/A

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1) Login with any user
2) Create a new project
3) Create a new sample inside this project
4) Upload 2 single end files to sample
5) In the rails console get the attachments for this sample and note their ids
6) Set a variable such as `u` to the user from above
7) Set a variable such as `s` to the sample from above
8) Call `Attachments::ConcatenationService.new(u, s, {attachment_ids: IDS_FROM_STEP_5_IN_ARRAY}).execute`
9) In the browser refresh the page and you should see the newly concatenated file, and the originals as well since we did not set the `delete_originals` flag
10) Remove the concatenated file from the sample and try step 8 again in rails console but this time pass in the `delete_originals: true`
11) Refresh the browser and you should see the newly concatenated file and the originals should be deleted
12) Try out the steps from above using paired-end files
13) Attempt to do the above steps using a `user` which doesn't have permissions to update the sample in a project, you should receive a `Not Authorized` error

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
